File: copasi/bindings/python/unittests/Test_CreateSimpleModel.py
# -*- coding: utf-8 -*-
# Copyright (C) 2019 - 2020 by Pedro Mendes, Rector and Visitors of the 
# University of Virginia, University of Heidelberg, and University 
# of Connecticut School of Medicine. 
# All rights reserved. 

# Copyright (C) 2017 - 2018 by Pedro Mendes, Virginia Tech Intellectual 
# Properties, Inc., University of Heidelberg, and University of 
# of Connecticut School of Medicine. 
# All rights reserved. 

# Copyright (C) 2010 - 2016 by Pedro Mendes, Virginia Tech Intellectual 
# Properties, Inc., University of Heidelberg, and The University 
# of Manchester. 
# All rights reserved. 

# Copyright (C) 2008 - 2009 by Pedro Mendes, Virginia Tech Intellectual 
# Properties, Inc., EML Research, gGmbH, University of Heidelberg, 
# and The University of Manchester. 
# All rights reserved. 

# Copyright (C) 2006 - 2007 by Pedro Mendes, Virginia Tech Intellectual 
# Properties, Inc. and EML Research, gGmbH. 
# All rights reserved. 

import COPASI
import unittest
from types import *

def createModel():
    datamodel=COPASI.CRootContainer.addDatamodel()
    model=datamodel.getModel()
    model.setVolumeUnit(COPASI.CUnit.fl)
    model.setTimeUnit(COPASI.CUnit.s)
    model.setQuantityUnit(COPASI.CUnit.fMol)
    comp=model.createCompartment("CompartmentA")
    A=model.createMetabolite("A",comp.getObjectName())
    B=model.createMetabolite("B",comp.getObjectName())
    react=model.createReaction("Decay_1")
    react.addSubstrate(A.getKey())
    react.addProduct(B.getKey())
    react.setReversible(False)
    react.setFunction("Mass action (irreversible)")
    mapping=COPASI.DataObjectVector()
    mapping.append(react.getChemEq().getSubstrate(0).getMetabolite())
    react.setParameterObjects(react.getFunction().getVariables().getParameter(1).getObjectName(),mapping);
    model.compileIfNecessary()

    A.setInitialConcentration(2.0e-4)    
    B.setInitialConcentration(0.0)
    react.setParameterValue("k1",0.5)
    changedObjects=COPASI.ObjectStdVector()
    changedObjects.push_back(comp.getObject(COPASI.CCommonName("Reference=InitialVolume")))
    changedObjects.push_back(A.getObject(COPASI.CCommonName("Reference=InitialConcentration")))
    changedObjects.push_back(B.getObject(COPASI.CCommonName("Reference=InitialConcentration")))
    model.updateInitialValues(changedObjects)

    return datamodel

def extendModel(datamodel):
    model=datamodel.getModel()
    metab=model.createMetabolite("C",model.getCompartment(0).getObjectName())
    react=model.createReaction("Decay_2")
    react.addSubstrate(model.getMetabolite(1).getKey())
    react.addProduct(metab.getKey())
    react.setReversible(False)
    react.setFunction("Mass action (irreversible)")
    mapping=COPASI.DataObjectVector()
    mapping.append(react.getChemEq().getSubstrate(0).getMetabolite())
    react.setParameterObjects(react.getFunction().getVariables().getParameter(1).getObjectName(),mapping);
    model.compileIfNecessary()
    metab.setInitialConcentration(0.0)
    react.getParameters().getParameter(0).setValue(0.1)
    changedObjects=COPASI.ObjectStdVector()
    changedObjects.push_back(metab.getObject(COPASI.CCommonName("Reference=InitialConcentration")))
    changedObjects.push_back(react.getParameters().getParameter(0).getObject(COPASI.CCommonName("Reference=Value")))
    model.updateInitialValues(changedObjects);


class Test_CreateSimpleModel(unittest.TestCase):
   def setUp(self):
    self.datamodel=createModel()
    self.model=self.datamodel.getModel()

   def test_createModel(self):
     self.assert_(self.model!=None)
     self.assert_(self.model.__class__==COPASI.CModel)
     self.assert_(self.model.getCompartments().size()==1)
     self.assert_(self.model.getCompartment(0).getObjectName()=="CompartmentA")
     self.assert_(self.model.getMetabolites().size()==2)
     self.assert_(self.model.getMetabolite(0).getObjectName()=="A")
     self.assertEquals(self.model.getMetabolite(0).getInitialConcentration(),2.0e-4)
     self.assert_(self.model.getMetabolite(1).getObjectName()=="B")
     self.assert_(self.model.getMetabolite(1).getInitialValue()==0.0)
     self.assert_(self.model.getReactions().size()==1)
     self.assert_(self.model.getReaction(0).getObjectName()=="Decay_1")
     self.assert_(self.model.getReaction(0).isReversible()==False)
     self.assert_(self.model.getReaction(0).getChemEq().getSubstrates().size()==1)
     self.assert_(self.model.getReaction(0).getChemEq().getSubstrate(0).getMetabolite().getObjectName()=="A")
     self.assert_(self.model.getReaction(0).getChemEq().getSubstrate(0).getMultiplicity()==1.0)
     self.assert_(self.model.getReaction(0).getChemEq().getProducts().size()==1)
     self.assert_(self.model.getReaction(0).getChemEq().getProduct(0).getMetabolite().getObjectName()=="B")
     self.assert_(self.model.getReaction(0).getChemEq().getProduct(0).getMultiplicity()==1.0)
     self.assert_(self.model.getReaction(0).getChemEq().getModifiers().size()==0)
     self.assert_(self.model.getReaction(0).getFunction()!=None)
     self.assert_(self.model.getReaction(0).getFunction().getObjectName()=="Mass action (irreversible)")
     self.assert_(self.model.getReaction(0).getParameters().size()==1)
     self.assert_(self.model.getReaction(0).getParameters().getParameter(0).getObjectName()=="k1")
     self.assert_(self.model.getReaction(0).getParameters().getParameter(0).getValue()==0.5)

   def test_extendModel(self):
     extendModel(self.datamodel) 
     self.assert_(self.model.getMetabolites().size()==3)
     self.assert_(self.model.getMetabolite(2).getObjectName()=="C")
     self.assertEquals(self.model.getMetabolite(2).getInitialValue(),0.0)
     self.assert_(self.model.getReactions().size()==2)
     self.assert_(self.model.getReaction(1).getObjectName()=="Decay_2")
     self.assert_(self.model.getReaction(1).isReversible()==False)
     self.assert_(self.model.getReaction(1).getChemEq().getSubstrates().size()==1)
     self.assert_(self.model.getReaction(1).getChemEq().getSubstrate(0).getMetabolite().getObjectName()=="B")
     self.assert_(self.model.getReaction(1).getChemEq().getSubstrate(0).getMultiplicity()==1.0)
     self.assert_(self.model.getReaction(1).getChemEq().getProducts().size()==1)
     self.assert_(self.model.getReaction(1).getChemEq().getProduct(0).getMetabolite().getObjectName()=="C")
     self.assert_(self.model.getReaction(1).getChemEq().getProduct(0).getMultiplicity()==1.0)
     self.assert_(self.model.getReaction(1).getChemEq().getModifiers().size()==0)
     self.assert_(self.model.getReaction(1).getFunction()!=None)
     self.assert_(self.model.getReaction(1).getFunction().getObjectName()=="Mass action (irreversible)")
     self.assert_(self.model.getReaction(1).getParameters().size()==1)
     self.assert_(self.model.getReaction(1).getParameters().getParameter(0).getObjectName()=="k1")
     self.assert_(self.model.getReaction(1).getParameters().getParameter(0).getValue()==0.1)

def suite():
  tests=[
          'test_createModel'
         ,'test_extendModel'
        ]
  return unittest.TestSuite(map(Test_CreateSimpleModel,tests))

if(__name__ == '__main__'):
    unittest.TextTestRunner(verbosity=2).run(suite())

 

File: copasi/bindings/java/unittests/Test_CreateSimpleModel.java
// Copyright (C) 2017 by Pedro Mendes, Virginia Tech Intellectual 
// Properties, Inc., University of Heidelberg, and University of 
// of Connecticut School of Medicine. 
// All rights reserved. 

// Copyright (C) 2010 - 2016 by Pedro Mendes, Virginia Tech Intellectual 
// Properties, Inc., University of Heidelberg, and The University 
// of Manchester. 
// All rights reserved. 

// Copyright (C) 2008 - 2009 by Pedro Mendes, Virginia Tech Intellectual 
// Properties, Inc., EML Research, gGmbH, University of Heidelberg, 
// and The University of Manchester. 
// All rights reserved. 

// Copyright (C) 2006 - 2007 by Pedro Mendes, Virginia Tech Intellectual 
// Properties, Inc. and EML Research, gGmbH. 
// All rights reserved. 

package org.COPASI.unittests;

import org.COPASI.*;
import junit.framework.*;

public class Test_CreateSimpleModel extends TestCase
{

   CDataModel mDataModel;
   CModel model;

   public Test_CreateSimpleModel(String name)
   {
    super(name);
   }

   public static CModel createModel(CDataModel dataModel)
   {
    try
    {
      dataModel.newModel();
    }
    catch(Exception e)
    {
        return null;
    }
    CModel model=dataModel.getModel();
    model.setVolumeUnit(CModel.fl);
    model.setTimeUnit(CModel.s);
    model.setQuantityUnit(CModel.fMol);
    CCompartment comp=model.createCompartment("CompartmentA");
    CMetab A=model.createMetabolite("A",comp.getObjectName());
    A.setInitialConcentration(2.0e-4);
    CMetab B=model.createMetabolite("B",comp.getObjectName());
    B.setInitialConcentration(0.0);
    CReaction react=model.createReaction("Decay_1");
    react.addSubstrate(A.getKey());
    react.addProduct(B.getKey());
    react.setReversible(false);
    react.setFunction("Mass action (irreversible)");
    react.setParameterValue("k1",0.5);
    StringStdVector mapping=new StringStdVector();
    mapping.add(react.getChemEq().getSubstrate(0).getMetabolite().getKey());
    react.setParameterMappingVector(react.getFunction().getVariables().getParameter(1).getObjectName(),mapping);
    model.compileIfNecessary();
    ObjectStdVector changedObjects=new ObjectStdVector();
    changedObjects.add(comp.getObject(new CCommonName("Reference=InitialVolume")));
    changedObjects.add(A.getObject(new CCommonName("Reference=InitialConcentration")));
    changedObjects.add(B.getObject(new CCommonName("Reference=InitialConcentration")));
    changedObjects.add(react.getParameters().getParameter(0).getObject(new CCommonName("Reference=Value")));
    model.updateInitialValues(changedObjects);
    return model;
   }

  public static void extendModel(CModel model)
  {
      CMetab metab=model.createMetabolite("C",model.getCompartment(0).getObjectName());
      metab.setInitialConcentration(0.0);
      CReaction react=model.createReaction("Decay_2");
      react.addSubstrate(model.getMetabolite(1).getKey());
      react.addProduct(metab.getKey());
      react.setReversible(false);
      react.setFunction("Mass action (irreversible)");
      react.getParameters().getParameter(0).setDblValue(0.1);
      StringStdVector mapping=new StringStdVector();
      mapping.add(react.getChemEq().getSubstrate(0).getMetabolite().getKey());
      react.setParameterMappingVector(react.getFunction().getVariables().getParameter(1).getObjectName(),mapping);;
      model.compileIfNecessary();
      ObjectStdVector changedObjects=new ObjectStdVector();
      changedObjects.add(metab.getObject(new CCommonName("Reference=InitialConcentration")));
      changedObjects.add(react.getParameters().getParameter(0).getObject(new CCommonName("Reference=Value")));
      model.updateInitialValues(changedObjects);
  }

   
  public void setUp()
  {
    mDataModel=CRootContainer.addDatamodel();
    this.model=this.createModel(mDataModel);
  }

  public void test_createModel()
  {
     assertFalse(this.model==null);
     assertTrue(this.model.getClass().getName().equals("org.COPASI.CModel"));
     assertTrue(this.model.getCompartments().size()==1);
     assertTrue(this.model.getCompartment(0).getObjectName().equals("CompartmentA"));
     assertTrue(this.model.getMetabolites().size()==2);
     assertTrue(this.model.getMetabolite(0).getObjectName().equals("A"));
     assertTrue(this.model.getMetabolite(0).getInitialConcentration()==2.0e-4);
     assertTrue(this.model.getMetabolite(1).getObjectName().equals("B"));
     assertTrue(this.model.getMetabolite(1).getInitialValue()==0.0);
     assertTrue(this.model.getReactions().size()==1);
     assertTrue(this.model.getReaction(0).getObjectName().equals("Decay_1"));
     assertTrue(this.model.getReaction(0).isReversible()==false);
     assertTrue(this.model.getReaction(0).getChemEq().getSubstrates().size()==1);
     assertTrue(this.model.getReaction(0).getChemEq().getSubstrate(0).getMetabolite().getObjectName().equals("A"));
     assertTrue(this.model.getReaction(0).getChemEq().getSubstrate(0).getMultiplicity()==1.0);
     assertTrue(this.model.getReaction(0).getChemEq().getProducts().size()==1);
     assertTrue(this.model.getReaction(0).getChemEq().getProduct(0).getMetabolite().getObjectName().equals("B"));
     assertTrue(this.model.getReaction(0).getChemEq().getProduct(0).getMultiplicity()==1.0);
     assertTrue(this.model.getReaction(0).getChemEq().getModifiers().size()==0);
     assertFalse(this.model.getReaction(0).getFunction()==null);
     assertTrue(this.model.getReaction(0).getFunction().getObjectName().equals("Mass action (irreversible)"));
     assertTrue(this.model.getReaction(0).getParameters().size()==1);
     assertTrue(this.model.getReaction(0).getParameters().getParameter(0).getObjectName().equals("k1"));
     assertTrue(this.model.getReaction(0).getParameters().getParameter(0).getDblValue()==0.5);
  }

  public void test_extendModel()
  {
     this.extendModel(this.model); ;
     assertTrue(this.model.getMetabolites().size()==3);
     assertTrue(this.model.getMetabolite(2).getObjectName().equals("C"));
     assertTrue(this.model.getMetabolite(2).getInitialValue()==0.0);
     assertTrue(this.model.getReactions().size()==2);
     assertTrue(this.model.getReaction(1).getObjectName().equals("Decay_2"));
     assertTrue(this.model.getReaction(1).isReversible()==false);
     assertTrue(this.model.getReaction(1).getChemEq().getSubstrates().size()==1);
     assertTrue(this.model.getReaction(1).getChemEq().getSubstrate(0).getMetabolite().getObjectName().equals("B"));
     assertTrue(this.model.getReaction(1).getChemEq().getSubstrate(0).getMultiplicity()==1.0);
     assertTrue(this.model.getReaction(1).getChemEq().getProducts().size()==1);
     assertTrue(this.model.getReaction(1).getChemEq().getProduct(0).getMetabolite().getObjectName().equals("C"));
     assertTrue(this.model.getReaction(1).getChemEq().getProduct(0).getMultiplicity()==1.0);
     assertTrue(this.model.getReaction(1).getChemEq().getModifiers().size()==0);
     assertFalse(this.model.getReaction(1).getFunction()==null);
     assertTrue(this.model.getReaction(1).getFunction().getObjectName().equals("Mass action (irreversible)"));
     assertTrue(this.model.getReaction(1).getParameters().size()==1);
     assertTrue(this.model.getReaction(1).getParameters().getParameter(0).getObjectName().equals("k1"));
     assertTrue(this.model.getReaction(1).getParameters().getParameter(0).getDblValue()==0.1);
  }
  

  public static void main(String[] args) {
    junit.textui.TestRunner.run(Test_CreateSimpleModel.class);
  }

}
